src/plugins/MediaManager.js

```diff
@@ -0,0 +1,345 @@
+'use strict';
+
+const log = require('loglevel');
+log.setDefaultLevel('debug');
+
+module.exports = class MediaManager {
+    /**
+     * Constructor to this MediaManager class
+     * @param {DeviceRenderer} instance root instance
+     * @param {boolean} videoWithMicrophone flag indicating wether the video must always be played with the microphone.
+     *  This flag is used when we don't want to split audio&video and instead want the audio bundled with the video in the same stream.
+     *  Defaults to the microphone flag in the instance options
+     * @param {number} videoWidth Maximum video width. Defaults to 1280
+     * @param {number} videoHeight Maximum video height. Defaults to 720
+     */
+    constructor(instance, videoWithMicrophone = instance.options.microphone, videoWidth = 1280, videoHeight = 720) {
+        this.instance = instance;
+        this.instance.mediaManager = this;
+
+        if (!navigator.mediaDevices || !this.setupPermissions()) {
```

Comment:
no need to await `this.setupPermissions()` ?

Reply:
Yes indeed, but correct me if i'm wrong, constructors cannot be async. 
We can use this.setupPermissions.then(()=>{}) but instance props (this.instance.mediaEventsEnabled, ...) may be initialized after they are used

I guess even if setupPermissions return false, since it's return a promise so !this.setupPermissions will be always true, the test pass.

missed this one :$

Reply:
> Yes indeed, but correct me if i'm wrong, constructors cannot be async.
I am far from fluent in javascript, so I really don't know. Kinda makes sense though to avoid a promise on a futur object.

Reply:
Is it better now?